makeReducedConfig.py

```diff
@@ -0,0 +1,271 @@
+                if modname not in self.ref_modules:
+                    self.ref_modules[modname] = line
+
+    def addRefMemories(self, fname):
+        with open(fname, "r") as f:
+            for line in f:
+                memname = line.split()[1].strip()
+                if memname not in self.ref_memories:
+                    self.ref_memories[memname] = line
+
+    def addNode(self, n):
+        if n.name not in self.nodes:
+            self.nodes[n.name] = n
+        else:
+            for c in n.connections:
+                self.nodes[n.name].addConnection(c)
+
+    def addConnection(self, c):
+        for c_ref in self.connections:
+            if c.isSame(c_ref): return
+        self.connections.append(c)
+
+    # Load in a project from a wires file
+    def loadProject(self, fname):
+        with open(fname, "r") as f:
+            for line in f:
+
+                # Get raw names from wires file
+                mem = line.split("input=>")[0].strip()
+                str_in = line.split("input=>")[1].split("output=>")[0].strip()
+                str_out = line.split("output=>")[1].strip()
+
+                # Create nodes and extensions from strings
+                node_in = node(str_in.split(".")[0])
+                in_ext = str_in.split(".")[1] if "." in str_in else ""
+                node_out = node(str_out.split(".")[0])
+                out_ext = str_out.split(".")[1] if "." in str_out else ""
+
+                # Add connection and nodes to project
+                new_c = connection(mem, node_in, node_out, in_ext, out_ext)
+                self.addConnection(new_c)
+                node_in.addConnection(new_c)
+                node_out.addConnection(new_c)
+                self.addNode(node_in)
+                self.addNode(node_out)
+
+    # Save a project to a wires file
+    def saveProject(self, fname, ref_proj):
+        with open(fname, "w") as f:
+            # Loop over full wire project to preserve ordering
+            for c in ref_proj.connections:
+                for c1 in self.connections:
+                    if c1.memory == c.memory:
+                        f.write(c.printConnection()+"\n")
+                        break
+
+    def saveModules(self, fname):
+        with open(fname, "w") as f:
+            # Loop over reference modules file to preserve ordering
+            for n in self.ref_modules:
+                if n in self.nodes:
+                    f.write(self.ref_modules[n])
+
+    def saveMemories(self, fname):
+        with open(fname, "w") as f:
+            # Loop over reference memories file to preserve ordering
+            for m in self.ref_memories:
+                for c in self.connections:
+                    if c.memory==m:
+                        f.write(self.ref_memories[m])
+                        break
+
+    def addTC(self, tc, layers, ref_p):
+        tcs = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L"]
```

Comment:
Please add comment explaining meaning of tcs, l1phis & lxphis.

Reply:
I've a stupid question. I'm confused by what this script should be doing. e.g. If I type "grep ME_L3PHIB16 wires.dat", I see that ME_L3PHIB16 reads one VMSME & one VMPROJ memory. But if I do the same using reduced_wires.dat, I see that it only reads the VMPROJ memory, but no VMSME memory. How will ME_L3PHIB16 work correctly without this missing memory? It can't even take a default value of zero, since it should contain the stubs to be matched to the track projections.

Reply:
I believe I've resolved this -- so initially I'd had this set up so it essentially only looks for inputs of inputs to whatever node you give as the place to start building the slice, and outputs of outputs. This leaves off any memories that are just inputs to some other node but not part of some linear chain. I've updated the script to check for inputs and outputs to all nodes that get added, and it will add anything that fits within the phi constraints that I imposed. I'm still learning the structure of this, so your checks are appreciated!

With this change, your first example looks good. I don't quite understand your second example (did you copy paste the same module again?) so I can't spot check that. I think with this change there's very little chance I'm now missing something that should be there -- I'd really only be worried that I've expanded the slice more than we intended because of some connections we don't actually want to duplicate.

Reply:
I still see, for example, that PR_L3PHIB reads 9 TPROJ memories in wires.dat, but only 1 TPROJ memory in reduced_wires.dat. This proc. module will be unhappy with 8 missing inputs. 
Also, reduced_processingmodules.dat contains 44 MatchEngines, so this is not a simple chain. I remain very uncertain what the goal is here. @pwittich can you please clarify?

Reply:
If your chain instantiates PR_L3PHIB, then I believe that all TPROJ memories read by this module must be included in reduced_wires.dat, since otherwise the firmware will not compile. I guess the plan is that the content of TPROJ_L1L2F_L3PHIB will be provided by the TC_L1L2F in the chain. and the content of all the other TPROJ connected to PR_L3PHIB will be taken from C++ emulation.
In general, I believe any module that you include include in reduced_wires.dat must include there the connections to the same input and output memories that it has in wires.dat. e.g. reduced_wires.dat should record that PR_L3PHIB is connected to TPROJ_L1L2E_L3PHIB, but should record the latter memory as having no input, to make clear that it comes from emulation.
https://indico.cern.ch/event/1017998/contributions/4272238/attachments/2208553/3737293/dps_20210315.pdf is the target chain.
@pwittich or @aehart could you confirm please.

Reply:
> If your chain instantiates PR_L3PHIB, then I believe that all TPROJ memories read by this module must be included in reduced_wires.dat, since otherwise the firmware will not compile. I guess the plan is that the content of TPROJ_L1L2F_L3PHIB will be provided by the TC_L1L2F in the chain. and the content of all the other TPROJ connected to PR_L3PHIB will be taken from C++ emulation.
> In general, I believe any module that you include include in reduced_wires.dat must include there the connections to the same input and output memories that it has in wires.dat. e.g. reduced_wires.dat should record that PR_L3PHIB is connected to TPROJ_L1L2E_L3PHIB, but should record the latter memory as having no input, to make clear that it comes from emulation.
> https://indico.cern.ch/event/1017998/contributions/4272238/attachments/2208553/3737293/dps_20210315.pdf is the target chain.
> @pwittich or @aehart or @aryd could you confirm please.

From today's HLS meeting discussion, my comment above is incorrect. However, every proc. module should have at least one input & one output, except if they are at the beginning/end of chain.